### ISSUE TITLE
Enhance todo matching to include loose matching fallback

### ISSUE THREAD

**Issue by scalvert** [ember-template-lint/ember-template-lint#1872]
Currently the behavior of the todo feature is that it attempts to match a violation with an existing todo file [using a hash of the file contents](https://github.com/ember-template-lint/ember-template-lint-todo-utils/blob/master/src/io.ts#L330). This works well for a number of cases, but falls apart on others, specifically

1. When a user edits a file with an existing todo, and through that editing changes the line numbers associated with that todo such that using the todo's file hash no longer matches (line and column are part of that hash)
2. Files are moved to another location (less severe, since the todo will just change to whatever severity is currently configured in the app, and the tool will correctly signal that a todo file is 'stale' and can be removed with `--fix`)

The primary issue we'd like to address is 1, which can degrade DX. To address this, we propose the following changes to the matching algorithm:

1. First attempt to match using the file hash (same as current implementation)
2. Fall back to generating a hash on the same file contents _using the violation's source instead of line and column_. This would allow us to attempt to loosely match the violation, preserving the todo.

This wouldn't be a breaking change, but would merely be an enhancement of the current functionality.

cc/ @rwjblue

**Comment by rwjblue** [comment]
Yep, this looks like a good summary of our conversation. Thanks for writing it up!

**Comment by MelSumner** [comment]
Started to work on this: https://github.com/ember-template-lint/ember-template-lint/issues/1872

**Comment by MelSumner** [comment]
@rwjblue @scalvert trying to capture the proposed solution criteria. 

## Summary of Solution

1. Check to see if the engine is correct
2. Check to see if the issue is still valid (not stale or expired)
3. Check to see if the hash has an exact match (so line/column is a match)
4. Check to see if the rule ID is a match
5. Check to see if the source (or checksum) is the same (TBD)

## Requirements

- The solution should not include any file name/hash/line & column changes 
- The solution should not conflict with `--clean-todo` functionality (or any existing functionality)

## Edge Cases & Remaining Questions

- if a developer opens a todo's `.json` file, and was to go to a template and look for the issue at the specified line and column, they would be confused if the information wasn't accurate 
- why wouldn't `yarn lint:hbs --clean-todo && yarn lint:hbs --update-todo` be sufficient to achieve the same outcome? Yes, it would change the created/warn/error dates (effectively resetting the clock) but there are so many workarounds for that already that it seems inconsistent to try to prevent that now. 
- is there a way to exclude the .lint-todo folder from search results by default?

**Comment by scalvert** [comment]
I think this partly captures it, but not the algorithm we spoke about. Let's try to capture that, as it's the most important detail that's missing from the above.

**Comment by MelSumner** [comment]
> I think this partly captures it, but not the algorithm we spoke about. Let's try to capture that, as it's the most important detail that's missing from the above.

@scalvert I didn't put some of the super-specific, function-related details in here (due to prior feedback that I didn't need to include all that), but I do have them in my notes for implementation. Can you edit the above comment to reflect what you'd like to see?

**Comment by scalvert** [comment]
Specifically, we should be calling out the algorithm that was discussed in the meeting yesterday, related to how we're matching sets of violations for a particular file. More specifically, we need to ensure that previous exact matches aren't attempted to be matched again when doing the fuzzy match pass.

> if a developer opens a todo's .json file, and was to go to a template and look for the issue at the specified line and column, they would be confused if the information wasn't accurate

Todo files aren't really stored or intended for human consumption. These values should be considered opaque, and it's not something that we should try to solve a problem for.

> why wouldn't yarn lint:hbs --clean-todo && yarn lint:hbs --update-todo be sufficient to achieve the same outcome? Yes, it would change the created/warn/error dates (effectively resetting the clock) but there are so many workarounds for that already that it seems inconsistent to try to prevent that now.

Two reasons:

1. The due dates would be reset
2. If a file contained an existing todo, a developer introduced a new error by editing the file, and in the process of introducing that new error they moved the lines corresponding to the first todo, running the commands you stated above would now delete the existing todo, and create two new todos. We don't necessarily want that, since it was only the existing todo that we wanted to target, not generating other new todos in the process.

> is there a way to exclude the .lint-todo folder from search results by default?

This would be IDE specific. Also, why would we want to do this?

> if we are trying to avoid writing to a disk in this solution, does that mean that we should be asking the user to resolve a fuzzy match?

I don't understand what you mean here. Could you explain what the exact workflow scenario you're referring to?
